### DISCUSSION TITLE
How does replication happen between two nodes of EMQX servers? Is it async or sync.

### DISCUSSION THREAD

**Discussion by shivamydv** [emqx/emqx#5872] [Q&A]
I am new to EMQX and I am trying to understand how does replication happen between nodes in EMQx and is it async or sync?

**Comment by k32** [comment]
The MQTT messages can be forwarded between the nodes of the cluster both synchronously and asynchronously: this is controlled by `rpc.mode` configuration parameter. By default it is set to `async`.
Topic metadata is replicated synchronously up until 4.3. In 5.0 we will add an option to make it async too.

**Reply by shivamydv** [reply]
Are there any docs that explains this process?

**Reply by k32** [reply]
Replication of the message payload and replication of the topic metadata are very different processes. Which one of them are you interested in?

**Reply by shivamydv** [reply]
Both of them

**Reply by k32** [reply]
1.You can find some information about message routing [here](https://docs.emqx.io/en/broker/v4.3/design/design.html#routing-layer-design)
2. All the metadata is stored in distributed [Mnesia](https://en.wikipedia.org/wiki/Mnesia) database, that takes care of replicating the data through out the cluster. Our approach to async Mnesia replication is documented  [here](https://github.com/emqx/mria/blob/main/doc/rlog_internals.md). This will be an optional feature in 5.0

**Reply by shivamydv** [reply]
Cool, Thanks!
